task.go

```diff
@@ -202,39 +208,95 @@ func Build(buildkitd *Buildkitd, outputsDir string, req Request) (Response, erro
 		}
 	}
 
+	if req.Config.OutputOCI {
+		err = loadOciImages(imagePaths, req)
+		if err != nil {
+			return Response{}, err
+		}
+	} else {
+		err = loadImages(imagePaths, req)
+		if err != nil {
+			return Response{}, err
+		}
+	}
+
+	return res, nil
+}
+
+func loadImages(imagePaths []string, req Request) error {
 	for _, imagePath := range imagePaths {
 		image, err := tarball.ImageFromPath(imagePath, nil)
 		if err != nil {
-			return Response{}, errors.Wrap(err, "open oci image")
+			return errors.Wrap(err, "open oci image")
 		}
 
 		outputDir := filepath.Dir(imagePath)
 
-		err = writeDigest(outputDir, image)
+		m, err := image.Manifest()
 		if err != nil {
-			return Response{}, err
+			return errors.Wrap(err, "get image manifest")
+		}
+
+		err = writeDigest(outputDir, m.Config.Digest)
+		if err != nil {
+			return err
 		}
 
 		if req.Config.UnpackRootfs {
-			err = unpackRootfs(outputDir, image, cfg)
+			err = unpackRootfs(outputDir, image, req.Config)
 			if err != nil {
-				return Response{}, errors.Wrap(err, "unpack rootfs")
+				return errors.Wrap(err, "unpack rootfs")
 			}
 		}
 	}
-
-	return res, nil
+	return nil
 }
 
-func writeDigest(dest string, image v1.Image) error {
-	digestPath := filepath.Join(dest, "digest")
+func loadOciImages(imagePaths []string, req Request) error {
+	for _, imagePath := range imagePaths {
+		_, err := os.Stat(imagePath)
+		if err != nil {
+			return errors.Wrapf(err, "image path %s not valid", imagePath)
+		}
 
-	manifest, err := image.Manifest()
-	if err != nil {
-		return errors.Wrap(err, "get image digest")
+		// go-containerregistry does not currently have support for loading a OCI formated
+		// image from a tarball, so we decompress it before doing anything.
+		logrus.Info("decompressing OCI image tar")
+		targetDir := filepath.Dir(imagePath)
+		imageDir := filepath.Join(targetDir, "image")
+		err = os.MkdirAll(imageDir, 0700)
+		if err != nil {
+			return errors.Wrapf(err, "unable to create image dir %s", imageDir)
+		}
+		run(os.Stdout, "tar", "-xvf", imagePath, "-C", imageDir)
+
+		l, err := layout.ImageIndexFromPath(imageDir)
+		if err != nil {
+			return errors.Wrapf(err, "failed to load %s as OCI layout", imagePath)
+		}
+
+		m, err := l.IndexManifest()
+		if err != nil {
+			return errors.Wrap(err, "error getting index manifest")
+		}
+
+		manifest := m.Manifests[0]
+
+		outputDir := filepath.Dir(imagePath)
+
+		err = writeDigest(outputDir, manifest.Digest)
```

Comment:
what is the difference of `manifest` between this line and line 240? 

Are both `manifest.Digest` and `m.Config.Digest` valid?

Reply:
Hey @xtremerui yep both are valid, it's just that there's a different way of getting the image digest for the ImageIndex as the ImageIndex can contain multiple images (per cpu architecture).